pydantic/_internal/_decorators.py

```diff
@@ -3,202 +3,274 @@
+    Callable,
+    Dict,
+    Generic,
+    Set,
+    Tuple,
+    TypeVar,
+    Union,
+    cast,
+    overload,
+)
 
-from pydantic_core.core_schema import GeneralValidatorFunction, JsonReturnTypes, ValidationInfo, WhenUsed
-from typing_extensions import Protocol
+from pydantic_core.core_schema import (
+    FieldValidationInfo,
+    FieldValidatorFunction,
+    FieldWrapValidatorFunction,
+    JsonReturnTypes,
+    ValidationInfo,
+    WhenUsed,
+)
+from typing_extensions import Protocol, TypeAlias
 
-from ..errors import PydanticUserError
+from pydantic._internal._repr import Representation
 
 if TYPE_CHECKING:
     from typing_extensions import Literal
 
-    from ..main import BaseModel
 
-__all__ = (
-    'FIELD_VALIDATOR_TAG',
-    'ROOT_VALIDATOR_TAG',
-    'Validator',
-    'ValidationFunctions',
-    'SerializationFunctions',
-    'prepare_validator_decorator',
-    'prepare_serializer_decorator',
-)
 FIELD_VALIDATOR_TAG = '_field_validator'
 ROOT_VALIDATOR_TAG = '_root_validator'
 
 FIELD_SERIALIZER_TAG = '_field_serializer'
 
 
-class Validator:
+class ValidatorDecoratorInfo(Representation):
     """
-    Store information about field and root validators.
+    A container for data from `@validator` so that we can access it
+    while building the pydantic-core schema.
     """
 
-    __slots__ = 'function', 'mode', 'sub_path', 'check_fields', 'is_field_validator'
+    def __init__(
+        self,
+        *,
+        fields: tuple[str, ...],
+        # pre=True/False in v1 should be converted to mode='before'/'after' in v2
+        mode: Literal['before', 'after'],
+        check_fields: bool | None,
+        each_item: bool,
+    ) -> None:
+        """
+        :param mode: the pydantic-core validator mode.
+        :param mode: Not yet supported.
+        :param check_fields: whether to check that the fields actually exist on the model.
+        """
+        self.fields = fields
+        self.mode = mode
+        self.check_fields = check_fields
+        self.each_item = each_item
+
+
+class FieldValidatorDecoratorInfo(Representation):
+    """
+    A container for data from `@field_validator` so that we can access it
+    while building the pydantic-core schema.
+    """
 
     def __init__(
         self,
         *,
+        fields: tuple[str, ...],
         mode: Literal['before', 'after', 'wrap', 'plain'],
-        is_field_validator: bool,
-        sub_path: tuple[str | int, ...] | None = None,
-        check_fields: bool | None = None,
-    ):
-        # function is set later after the class is created and functions are bound
-        self.function: Callable[..., Any] | None = None
+        sub_path: tuple[str | int, ...] | None,
+        check_fields: bool | None,
+    ) -> None:
+        """
+        :param mode: the pydantic-core validator mode.
+        :param type: either 'unbound' or 'field' indicating if this validator should have
+            access to the model itself.
+        :param mode: Not yet supported.
```

Comment:
This looks like a typo since it's repeated twice

Reply:
thanks, fixed